From 75fc74b7e6fc224188513f6f9d967cb05c1ca5d0 Mon Sep 17 00:00:00 2001
From: Hauke Mehrtens <hauke.mehrtens@intel.com>
Date: Tue, 13 Oct 2020 16:02:17 +0200
Subject: [PATCH] PONRTSYS-8120: Fix possible resource leak

This simplifies the code and fixes a possible resource leak in an error
path.
---
 drivers/net/datapath/pon_qos/pon_qos_tc_qmap.c | 16 +++++++---------
 1 file changed, 7 insertions(+), 9 deletions(-)

diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_qmap.c b/drivers/net/datapath/pon_qos/pon_qos_tc_qmap.c
index 675456686cf6..caed1b0b9e26 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_qmap.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_qmap.c
@@ -71,6 +71,7 @@ static int pon_qos_parse_tc_flower(struct tc_cls_flower_offload *f,
 				BIT(FLOW_DISSECTOR_KEY_PORTS))) {
 		pr_debug("%s: Unsupported key used: 0x%x\n", __func__,
 			 f->dissector->used_keys);
+		kfree(*flt);
 		return -EINVAL;
 	}
 	pr_debug("%s: Supported key used: 0x%x\n", __func__,
@@ -121,15 +122,11 @@ int pon_qos_tc_map(struct net_device *dev, struct tc_cls_flower_offload *f)
 
 	ret = pon_qos_tc_flower_action_valid(f->exts);
 	if (ret < 0)
-		return ret;
+		goto err_put_indev;
 
 	ret = pon_qos_parse_tc_flower(f, &map);
-	if (ret < 0) {
-		if (map)
-			goto err;
-		else
-			return ret;
-	}
+	if (ret < 0)
+		goto err_put_indev;
 
 	map->dev = dev;
 	map->indev = indev;
@@ -152,9 +149,10 @@ int pon_qos_tc_map(struct net_device *dev, struct tc_cls_flower_offload *f)
 
 	return 0;
 err:
-	if (map->indev)
-		dev_put(map->indev);
 	kfree(map);
+err_put_indev:
+	if (indev)
+		dev_put(indev);
 	return ret;
 }
 EXPORT_SYMBOL(pon_qos_tc_map);
